data-analytics/dataflow-python-examples/dataflow_python_examples/data_generation_for_benchmarking.py

```diff
@@ -0,0 +1,485 @@
+        prior PCollection.
+
+        Attributes:
+            data_gen: A DataGenerator object defining the shape of the data should be generated by
+                      this DoFn.
+            n: An integer specifying how many records to generate.
+        """
+        self.data_gen = data_gen
+
+    # Helper function to get a single field dictionary from the schema for checking type and mode.
+
+    def get_field_dict(self, field_name):
+        return filter(lambda f: f[u'name'] == field_name,
+                      self.data_gen.schema)[0]
+
+    def sanity_check(self, record, fieldname):
+        """
+        This function ensures that the data is all of types that BigQuery expects.
+
+        Args:
+            record: A dictionary generated by faker_schema representing the candidate for a fake
+                    row in our BigQuery table
+            fieldname: name of field we are checking with this call.
+        """
+        # Create a Faker instance for individual parameterized random generation
+        # (ie. minimum date).
+        faker = Faker()
+
+        field = self.get_field_dict(fieldname)
+        # Below handles if the datatype got changed by the faker provider
+        if field[u'type'] == 'STRING' or field[u'type'].find('TIME') > -1:
+            record[fieldname] = unicode(record[fieldname])
+        elif field[u'type'].find('DATE') > -1:
+            # This implements the minimum/maximum date functionality
+            # and avoids regenerating a random date if already obeys min/max date.
+            if record[fieldname] < self.data_gen.min_date \
+                    or record[fieldname] > self.data_gen.max_date:
+                record[fieldname] = faker.date_between(self.data_gen.min_date,
+                                                       self.data_gen.max_date)
+            record[fieldname] = unicode(record[fieldname].strftime('%Y-%m-%d'))
+        elif field[u'type'] == 'INTEGER':
+            # This implements max and sign constraints
+            # and avoids regenerating a random integer if already obeys min/max integer.
+            if record[fieldname] > self.data_gen.max_int:
+                record[fieldname] = np.random.randint(0 if self.data_gen.only_pos
+                                                      else -1 * self.data_gen.max_int,
+                                                      self.data_gen.max_int)
+            if self.data_gen.only_pos and record[fieldname] < 0:
+                record[fieldname] = abs(record[fieldname])
+            record[fieldname] = int(record[fieldname])
+        elif field[u'type'] == 'FLOAT':
+            # This implements max and sign constraints
+            record[fieldname] = faker.pyfloat(math.log10(self.data_gen.max_float),
+                                              self.data_gen.float_precision,
+                                              self.data_gen.only_pos)
+            record[fieldname] = float(record[fieldname])
```

Comment:
This looks like it's throwing away the original value of `record[fieldname]`

Reply:
Adding check to see if the original value violates max_float, or only_pos.
```python
        elif field['type'] == 'FLOAT':
            # This implements max and sign constraints
            if (record[fieldname] > self.data_gen.max_float or
                    (record['fieldname'] < 0 and self.data_gen.only_pos)):
                record[fieldname] = faker.pyfloat(math.log10(self.data_gen.max_float),
                                                  self.data_gen.float_precision,
                                                  self.data_gen.only_pos)
            record[fieldname] = round(float(record[fieldname]), float_precision)
```